xarray/core/parallel.py

```diff
@@ -0,0 +1,333 @@
+
+    if not isinstance(args, Sequence):
+        raise TypeError("args must be a sequence (for example, a list or tuple).")
+
+    if kwargs is None:
+        kwargs = {}
+
+    elif not isinstance(kwargs, Mapping):
+        raise TypeError("kwargs must be a mapping (for example, a dict)")
+
+    if not dask.is_dask_collection(obj):
+        return func(obj, *args, **kwargs)
+
+    if isinstance(obj, DataArray):
+        # only using _to_temp_dataset would break
+        # func = lambda x: x.to_dataset()
+        # since that relies on preserving name.
+        if obj.name is None:
+            dataset = obj._to_temp_dataset()
+        else:
+            dataset = obj.to_dataset()
+        input_is_array = True
+    else:
+        dataset = obj
+        input_is_array = False
+
+    input_chunks = dataset.chunks
+
+    template = infer_template(
+        func, obj, *args, **kwargs
+    )  # type: Union[DataArray, Dataset]
+    if isinstance(template, DataArray):
+        result_is_array = True
+        template_name = template.name
+        template = template._to_temp_dataset()
+    elif isinstance(template, Dataset):
+        result_is_array = False
+    else:
+        raise TypeError(
+            "func output must be DataArray or Dataset; got %s" % type(template)
+        )
+
+    template_indexes = set(template.indexes)
+    dataset_indexes = set(dataset.indexes)
+    preserved_indexes = template_indexes.intersection(dataset_indexes)
+    new_indexes = set(template_indexes) - set(dataset_indexes)
+    indexes = {dim: dataset.indexes[dim] for dim in preserved_indexes}
+    indexes.update({k: template.indexes[k] for k in new_indexes})
+
+    graph = {}  # type: Dict[Any, Any]
+    gname = "%s-%s" % (dask.utils.funcname(func), dask.base.tokenize(dataset))
+
+    # map dims to list of chunk indexes
+    ichunk = {dim: range(len(chunks_v)) for dim, chunks_v in input_chunks.items()}
+    # mapping from chunk index to slice bounds
+    chunk_index_bounds = {
+        dim: np.cumsum((0,) + chunks_v) for dim, chunks_v in input_chunks.items()
+    }
+
+    # iterate over all possible chunk combinations
+    for v in itertools.product(*ichunk.values()):
+        chunk_index_dict = dict(zip(dataset.dims, v))
+
+        # this will become [[name1, variable1],
+        #                   [name2, variable2],
+        #                   ...]
+        # which is passed to dict and then to Dataset
+        data_vars = []
+        coords = []
+
+        for name, variable in dataset.variables.items():
+            # make a task that creates tuple of (dims, chunk)
+            if dask.is_dask_collection(variable.data):
+                # recursively index into dask_keys nested list to get chunk
+                chunk = variable.__dask_keys__()
+                for dim in variable.dims:
+                    chunk = chunk[chunk_index_dict[dim]]
+
+                chunk_variable_task = ("tuple-" + dask.base.tokenize(chunk),) + v
```

Comment:
You might keep the same name as above.  It's probably not necessary to re-tokenize everytihng.  There is also some benefit while debugging in all of them having the same hash, just so that we know what parts of the computation are related.

Reply:
thanks. now using the same name and replaced tokenize call with chunk[0]